Initial migration of site to 11ty

parlough opened the pull request:
I'm sorry that this ended up as such a large pull request, but that's mostly because it essentially touched every file with a small set of changes. See the different sections below for considerations for review or future reference. Please note that we'd like this to land despite it not being perfect, as then pain points can be identified and improved. Then I can also follow up with proper documentation as the implementation and syntax is stabilized. I'll be staying on top of any reported issue. If anything goes majorly wrong, we can always revert the change as well. Thanks for your time 💙

Resolves https://github.com/dart-lang/site-www/issues/5177
Resolves https://github.com/dart-lang/site-www/issues/3846
Resolves https://github.com/dart-lang/site-www/issues/5323
Fixes https://github.com/dart-lang/site-www/issues/4297
Fixes https://github.com/dart-lang/site-www/issues/4919
Closes https://github.com/dart-lang/site-www/issues/4631
Contributes to https://github.com/dart-lang/site-www/issues/4163

> [!TIP]
> Reviewing this in GitHub's UI might be slow. Consider cloning the branch locally.

**Changes potentially worth reviewing from a technical perspective are:**

- Implemented 11ty in `eleventy.config.js` with customizations in `/src/_11ty`.
  - `eleventy.config.js` configures the site build, enables used plugins, and connects the custom logic added in `/src/_11ty`. This will be greatly simplified and better documented in follow-up work.
- Replaced prettify, rouge, and custom code excerpter logic with custom markdown plugin.
  - You can find this at `/src/_11ty/plugins/highlight.js` _(super messy for now)_.
  - New custom theme is at `/src/_11ty/syntax/dash-light.js`.
  - The implementation of this needs a lot of cleanup and will be simplified greatly in follow-up work to rely on recent improvements in underlying highlighting package.
- Added a `build` and `serve` command to `dart run dart_site` (found in `/tool/dart_site/lib`).
- The deploy configs in `/cloud_build` have been migrated from using Docker to use Node to build the site. I reused the Docker image we built for firebase_tools, as that already has Node. Let me know if I shouldn't though :)
- Removed remnants of previous infra (Jekyll files, gem files, docker files, makefile, etc).
- All written docs and assets have been moved into `/content` to enable a separation from site tooling and configuration.

**Changes potentially worth reviewing or knowing about from a writer/contributor perspective:**

- **Major:** All doc content has been moved into `/content`. This allows site implementation to be separate from content, potentially enabling easier downstream localization as well.
- The pre-existing global variables are now defined in `/src/_data/site.yml`.
- Alerts/asides use a new syntax and have a design closer to other doc sites, search for `:::` for examples.
- Prettify doesn't exist anymore. Always use Markdown code blocks. They now always support highlighting with `[! !]`.
- Terminal code blocks now use `console` instead of `terminal`. Just textual ones use `plaintext` instead of `nocode`.
- Markdown code blocks now support specifying a title/filename with syntax like `title="main.dart"`. Various code blocks have been updated to use it instead of one-off solutions.
- Code blocks have a new design, with slightly bigger text, new colors, reduced padding, a more defined border, and a tag with the code block's language.
- The README is updated with the core setup requirements. Note that more comprehensive docs will come.
- To change the anchor of a header, a slightly modified `{:#new-id}` syntax is now used.
- The include syntax is slightly different, put quotes around the include filename and define arguments with colons: `{% include 'linter-rule-mention.md', rule:'unnecessary_getters_setters' %}`.

**Current issues:**

The following are some issues or regressions that I plan to accept in the short term to get this landed.

- Terminal/console code blocks do not have their formatted black background yet. They appear as normal code blocks.
- Incremental rebuild is slower than it should be due to a new bug in 11ty.

**Follow-up work:**

The following are some remaining tasks and improvements that I'd like to complete in follow-up so this can land and stretch its feet. These improvements can then be reviewed in isolation and incorporate what I've learned from deploying.

- Incorporate fix for 11ty's current incremental rebuild issue.
- Add back external link icons where it makes sense during static build.
- Reintroduce terminal code block's having a custom style.
- Standardize old and new liquid functions and filters to modern 11ty suggestions for consistent experience.
- Document the 11ty, liquid, and custom features we use that are relevant to site writers and contributors.
- Speed up sidenav and TOC generation. Potentially reduce needed dependencies while doing so.
- Clean up implementation of custom 11ty and Markdown plugins. Rely on new features in underlying syntax highlighting plugin rather than implementing our own form.
- Automate image optimization in production builds.
- Use TypeScript for custom 11ty functionality to avoid painfully adding JSDocs everywhere.
- Add back a devcontainer setup for users who want to a functioning workflow in the browser.
- Add skippable setup checks to `dart_site` tool
- Move the Cloud Build commands to `dart_site` tool

khanhnwin commented:
You're awesome, @parlough! I'm slowly but surely making my way through this PR! <3 Thank you for doing this!!

chalin commented on README.md:
```diff
@@ -86,155 +87,142 @@ _choose one_ of the following submodule-cloning techniques:
-   > For details, see [#1363](https://github.com/flutter/website/issues/1363).
 
-6. Make your changes to the local repo.
+   The port might be different if `4000` is taken.
+   If you need to view the generated fields,
+   they can be found in the `_site` directory.
```
```suggestion
   > [!NOTE]
   > The port might be different if `4000` is taken. For generated site files,
   > see the `_site` directory.
```

chalin commented on src/content/overview.md:
```diff
@@ -107,15 +107,15 @@ class Point {
+This example is running in an embedded [DartPad](/tools/dartpad).
+You can also
+<a href="{{site.dartpad}}/bc63d212c3252e44058ff76f34ef5730"
+target="_blank" rel="noopener">open this example in its own window</a>.
```
FYI, external links like these aren't displayed with an external-link icon anymore.

parlough replied:
Yeah I left this out on purpose to figure out as follow up.

The Google documentation style guide recently [updated its guidelines](https://developers.google.com/style/whats-new#november-8,-2023) to suggest [not adding external link icons](https://developers.google.com/style/link-text#write-link-text:~:text=Don%27t%20use%20an%20external%20link%20icon%20to%20indicate%20that%20the%20link%20goes%20to%20a%20different%20domain%20or%20server) and instead improve surrounding context. They explained that "Readers and writers have expressed confusion about the meaning and usage of this icon".

My initial thinking is that in most cases it's sufficient to explain the destination in the surrounding text/context. However, in UI elements like the sidenav, I feel external icons might still necessary.

chalin replied:
> The Google documentation style guide recently [updated its guidelines](https://developers.google.com/style/whats-new#november-8,-2023) to suggest [not adding external link icons](https://developers.google.com/style/link-text#write-link-text:~:text=Don%27t%20use%20an%20external%20link%20icon%20to%20indicate%20that%20the%20link%20goes%20to%20a%20different%20domain%20or%20server) and instead improve surrounding context. They explained that "Readers and writers have expressed confusion about the meaning and usage of this icon".

Oh! That is very good to know. Thanks for providing those links to the style guide.

chalin commented on src/content/overview.md:
```diff
@@ -107,15 +107,15 @@ class Point {
 
 
-## Dart: The libraries {#libraries}
+## Dart: The libraries {:#libraries}
```
I'm curious, is the `{:#...}` syntax required by 11ty?

parlough replied:
**Short explanation:** No, I configured it that way.

**Longer explanation:**

Compared to Jekyll, 11ty comes with a lot less in the box, but that also allows it to be more configurable.

The default Markdown package is `markdown-it` and this support comes from a plugin for that called [`markdown-it-attrs`](https://www.npmjs.com/package/markdown-it-attrs). It defaults to just `{` as the starting delimiter, but I [configured](https://github.com/dart-lang/site-www/blob/f4f49bf27ae8f3482e320b69e431c35ada6f2b6a/src/_11ty/plugins/markdown.js#L15-L16) it to start with `{:`. I had a reason at the time, but I am struggling to recall why now... If I remember, I'll document it, otherwise I might go back to the default. Thanks for asking!

chalin commented on src/content/effective-dart/style.md:
```diff
@@ -35,14 +35,13 @@ Identifiers come in three flavors in Dart.
 ### DO name types using `UpperCamelCase`
 
-{% include linter-rule-mention.md rule="camel_case_types" %}
+{% include 'linter-rule-mention.md', rule:'camel_case_types' %}
```
The linter rule lines are styled differently (maybe you know already?):

- Before:
  > <img width="350" alt="image" src="https://github.com/dart-lang/site-www/assets/4140793/90dc4c53-da0f-47e7-be33-90bb9f8980bb">


- After:
  > <img width="350" alt="image" src="https://github.com/dart-lang/site-www/assets/4140793/00499e5e-e365-4142-9bc1-cd8f3e3e0e8a">

parlough replied:
Good catch, I didn't notice this one!

I went ahead and updated the include to add the styling back and accept `rules` rather than `rule`, `rule1`, and `rule2`.

chalin reviewed (commented):
Oh my goodness! I'm floored 😲. This is a **_phenomenal upgrade_**! Well done @parlough! 🙌🏻

No joke that "GitHub's UI might be slow" 🙃

I've locally tested the setup, serve, and check instructions. Not only do they work like a charm, but the entire process was so quick! 🐎 

I've included a few tweaks to the README as comments, and noted a few differences in the rendered site. You might know about these differences already (and I understand that you'rre wanting to tackle the upgrade incrementally), but I thought I'd mention them just in case.

I happened to stumble upon this work because I was wondering about the status of the code-excerpt tools. In the opening comment you state "Replaced ... custom code excerpter logic with custom markdown plugin". But I get the impression that the code-excerpt tools are still being invoked via `tool/dart_site/lib/src/commands/refresh_excerpts.dart`, is that correct? 

Again, bravo for such an amazing job!

chalin commented:
> Markdown code blocks now support specifying a title/filename with syntax like title="main.dart". Various code blocks have been updated to use it instead of one-off solutions.

@parlough - I wanted to see some examples of the use of `title=...` but didn't find any. Can you point me to one? Thanks!

parlough commented:
👋 Hey Patrice @chalin! I really appreciate you taking a look! I hope you are doing well :)

> This is a phenomenal upgrade! Well done @parlough! 🙌🏻

I'm not sure I would say phenomenal with some of my 11ty and highlighting implementation 🙈, but I'm glad you think so! JS is at least a lot more familiar to all of us than Ruby.

For more context: The code block functionality ended up being the bulk of my mess and change here. Since I implemented it though, the syntax highlighting package I was using added a bunch of new functionality and configurability. Once this lands I'll try to use as much of that as I can. It looks like they consolidated with their upstream package and will release 1.0.0 soon ([shiki](https://github.com/shikijs/shiki)).

> I've locally tested the setup, serve, and check instructions. Not only do they work like a charm, but the entire process was so quick! 🐎

That's the goal!! Ruby, Jekyll, and Docker or some combination of them have been causing issues often and I think Googlers can't easily use Docker on their macs anymore. I would love all of this to be implemented in Dart, but for now, I think I can safely assume developers interested in contributing are more likely to have Node setup than Ruby.

I would also like to add a mechanism for using the `dart_site` tooling with a simple command from any nested directory. I haven't decided the best way though. Perhaps globally activating it with pub or a shell and ps script that you can add to your path?

> I've included a few tweaks to the README as comments, and noted a few differences in the rendered site. You might know about these differences already (and I understand that you'r wanting to tackle the upgrade incrementally), but I thought I'd mention them just in case.

I didn't realize all of these, so thank you! Some I'm planning on leaving for follow-up work to keep this PR from ballooning further.

> I happened to stumble upon this work because I was wondering about the status of the code-excerpt tools. But I get the impression that the code-excerpt tools are still ... is that correct?

They're mostly the same as you left them, outside of a few fixes and being invoked through that Dart script you mentioned. The changes in this PR are about not using the instructions in code block generation, so handling all Markdown code blocks the same. 

Based on how we've ended up using the tools though, I started working on a version of the tooling with less functionality that doesn't use `package:build_runner` in https://github.com/dart-lang/site-shared/pull/201. It's not done and needs to be tested, but I'd like to get back to it soon :D. If you're interested, I'd appreciate any feedback you have on that!

> @parlough - I wanted to see some examples of the use of title=... but didn't find any. Can you point me to one? Thanks!

There aren't too many usages yet, but there's a few in [`/effective-dart/usage#prefer-relative-import-paths`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/effective-dart/usage#prefer-relative-import-paths). Each linter rule page uses them as well: [`/tools/linter-rules/avoid_relative_lib_imports#usage`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/tools/linter-rules/avoid_relative_lib_imports#usage).

For line numbers, I think [`/tools/analysis`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/tools/analysis) might be the only usage so far.

---

A side note, thanks for your documentation and examples on updating Docsy and its downstream sites to Bootstrap 5! I really need to work on that migration soon and your work will be a super useful reference.

parlough commented on README.md:
```diff
@@ -2,69 +2,70 @@
+built with [Eleventy][] and hosted on [Firebase][].
 
-[We welcome contributions](CONTRIBUTING.md),
-and we're [first-timer friendly][first-timers]!
```
We are! But when I was updating the README I realized the label isn't there anymore, so I removed this link for now to avoid any confusion.

I have follow-up work to introduce more contributing docs for the site infra, syntax, and writing practices. Once that's done and this setup is finalized, I'll bring something back here so it's actually easy to get going as a "first-timer".

parlough commented on README.md:
```diff
@@ -86,155 +87,142 @@ _choose one_ of the following submodule-cloning techniques:
+   This command generates and serves the site on a
+   local port that's printed to your terminal.
 
 5. View your changes in the browser by navigating to `http://localhost:4000`.
```
Good idea, thanks!

parlough commented on README.md:
```diff
@@ -2,69 +2,70 @@
-  We use Docker for local dev, tests, and building the site.
-  Install it from https://docs.docker.com/get-docker/.
+- **Node.js**
+  The latest stable LTS release of Node.js is required to build the site.
```
I've never used nvm, but this seems like an easy win for those that do. Added! Is it popular enough that I should add a mention to the README?

Thanks :)

atsansone replied:
@parlough : Indeed it is.

khanhnwin commented on .github/workflows/test.yml:
```diff
@@ -57,25 +57,42 @@ jobs:
+      - uses: pnpm/action-setup@ebcfd6995dade4b0104ac774445cef8b3b4635b0
+        with:
+          version: 8
       - uses: actions/setup-node@b39b52d1213e96004bfcb1c61a8a6fa8ab84f3e8
```
For my own understand, what/why pnpm vs npm?

parlough replied:
I'm not super up to date, but a lot of the Node ecosystem has been standardizing towards pnpm. In testing both for this change, I also found its caching and node_modules handling to be much faster and friendlier. Friendlier in terms of switching between branches and IDEs better understanding what dependencies you have available.

Technically someone can use `npm` with this setup as well if they'd like, it's just slower.

khanhnwin commented on .github/workflows/test.yml:
```diff
@@ -57,25 +57,42 @@ jobs:
+      - name: Check for broken Markdown links
+        run: dart run dart_site check-link-references
       - name: Check internal site links are functional
         run: dart run dart_site check-links
```
dart_site references the tooling that's in tool/dart_site right?

parlough replied:
Yep! This will be consolidated under a simpler one word command in the future, like `dart_site check-links` or something.

khanhnwin replied:
PERFECT!

khanhnwin commented on cloud_build/deploy.yaml:
```diff
@@ -1,20 +1,16 @@
-        echo "Building the website using a makefile..."
-        make build
-        make write-prod-robots
   - name: gcr.io/flutter-dev-230821/firebase-ghcli
```
Where did the step to build the static files go? Is that the equivalent of 'npm run prod' now?

khanhnwin replied:
If so, could we make that "build_prod_site_files" or something that makes it clearer that we're building the files?

parlough replied:
Yep! `npm run build` builds the static site files for staging. I didn't need the earlier step as we don't use Docker now and the `gcr.io/flutter-dev-230821/firebase-ghcli` has Node available to use.

parlough replied:
> Could we make that "build_prod_site_files" or something ...

Could you specify what you're referring to? Do you mean the name of the `npm build` command?

khanhnwin commented on cloud_build/stage.yaml:
```diff
@@ -1,20 +1,16 @@
+        set -e
+        
+        npm install
+        npm run build
```
What's the difference between prod vs build?

parlough replied:
A production build minifies HTML, CSS, etc. It also has a `robots.txt` to allow the site to be indexed, since we don't want staged sites to be indexed.

This will eventually be more clear once I can use the `dart_site` tooling here and use those commands, but Dart isn't available in these Cloud Build workflows currently. So I added `npm build` and `npm prod` to use temporarily.

khanhnwin replied:
Thanks Parker :D

khanhnwin commented on examples/misc/lib/effective_dart/usage_bad.dart:
```diff
@@ -9,9 +9,7 @@
 // ignore_for_file: unchecked_use_of_nullable_value, unnecessary_library_directive
```
I'm not as familiar with the dart.dev codebase, but why is this examples directory at the root as opposed to src/?

parlough replied:
It's the same as `flutter/website`, but I guess I don't have a reason. They've just always been that way since I started working on the sites.

Perhaps it makes sense though, as the content isn't actually used in the site or by the site generation itself, but rather by the separate code excerpt and testing functionality.

Happy to discuss the examples directory more if you're interested or have any thoughts! :D

khanhnwin replied:
This seems to be out of scope for this project haha but worth discussing later on once migration is done!

khanhnwin commented on tool/dart_site/lib/src/commands/build.dart:
```diff
@@ -0,0 +1,42 @@
+  BuildSiteCommand() {
+    argParser.addFlag(
+      _releaseFlag,
+      defaultsTo: false,
```
This flag defaults to false

parlough replied:
Good point!

So the optional parameter for a flag that defaults to false, also defaults to false 🤣 

Do you mind if I keep these though? It's then easier for someone coding in these files to identify the flag defaults to `false` rather than `true` or anything else, especially if they're not super familiar with `package:args`.

khanhnwin replied:
Yeah, that's fair!

parlough commented on tool/dart_site/lib/src/commands/build.dart:
```diff
@@ -0,0 +1,42 @@
+
+  @override
+  Future<int> run() async {
+    final productionRelease = argResults.get<bool>(_releaseFlag, false);
```
Thanks for calling this out! I'll add some more context to the help message. But specifying `--release` enables html and css minification, production `robots.txt`, etc. Potentially other optimizations in the future that aren't necessary for development or make debugging harder.

Generally this command and flag (`dart_site build --release`) should be used `npm run prod`.

khanhnwin replied:
Gotcha! I wasn't sure if _releaseFlag was already defaulting to false and then we set another default on top of it here.

parlough replied:
OHH, sorry! You're right it technically defaults to `false` already but due to the design of `package:args` not having any static guarantees I added this `argResults.get<T>` method as a helper to add static typing to retrieving the arguments.

khanhnwin replied:
Gotcha! Makes sense!

khanhnwin commented on tool/dart_site/lib/dart_site.dart:
```diff
@@ -29,5 +31,7 @@ final class DartSiteCommandRunner extends CommandRunner<int> {
     addCommand(AnalyzeDartCommand());
```
Do we have a "man dart_site" type of doc anywhere that documents these commands? That would be super helpful!

parlough replied:
You can run `dart run dart_site --help` to get a list of each command with their documentation. Once I've simplified their usage I'll be including how to use these in task-based site infra documentation.

khanhnwin commented on AUTHORS:
```diff
@@ -13,3 +13,4 @@ Chris Buckett <chrisbuckett@gmail.com>
 Michael Haubenwallner <michael.haubenwallner@gmail.com>
 Victor Berchet <victor.berchet@gmail.com>
 Pradumna Saraf <pradumnasaraf@gmail.com>
+Parker Lougheed <parlough@gmail.com>
```
🎉

khanhnwin commented on README.md:
```diff
@@ -2,69 +2,70 @@
+For simple changes (such as style and typo fixes),
+you probably don't need to build this site.
+Often you can make changes using the GitHub UI.
+We can stage the changes automatically in your pull request.
```
Great call out!

parlough replied:
It was just an adjustment of what was there, but thanks :P

- While the new setup is easier, I still want to discourage developers looking to quickly contribute.

khanhnwin commented on README.md:
```diff
@@ -86,155 +87,144 @@ _choose one_ of the following submodule-cloning techniques:
+   
+   To instead view the source of generated site files,
+   check the `_site` directory.
 
```
is "fields" supposed to be "files" here?

parlough replied:
It is already "files", but yes :)

khanhnwin commented on README.md:
````diff
@@ -86,155 +87,144 @@ _choose one_ of the following submodule-cloning techniques:
-   > ```terminal
-   > $ make down && make clean && make up
-   > ```
+   The site should automatically rebuild on most changes, but if
+   something doesn't update, exit the process and rerun the command.
````
Are there "known issues" here that we can flag for folks?

khanhnwin replied:
alternatively, an issue link or something

parlough replied:
I'd like to see which issues come up naturally, but I'll leave a comment to open an issue. Good point!

khanhnwin commented on README.md:
```diff
@@ -86,155 +87,144 @@ _choose one_ of the following submodule-cloning techniques:
-
-1. Navigate to your PR on GitHub and update it with the location of
-  the staged version, the names of your reviewers, and so on.
+4. Navigate to your PR on GitHub and include the link of the staged version.
+   Do consider adding a reference to the commit you staged,
+   so that reviewers know if any further changes have been made.
```
doesn't our staging GitHub build handle this?

parlough replied:
Yes it does. This entire section is for if you'd like to stage on your own or you need to for some reason.

khanhnwin replied:
Can we perhaps add an [Optional] to the title intro section at the top title? Want to flag this is not required and folks should only do it if they want to for their own purposes.

khanhnwin commented on package.json:
```diff
@@ -1,13 +1,42 @@
   "version": "0.0.0",
-  "private": false,
-  "description": "Source code for https://dart.dev",
+  "private": true,
```
do we want to publish?

parlough replied:
Nope since the site isn't a package. Setting "private" to `true` ensures no one will accidentally try to do so hhh.

parlough commented on src/_layouts/none.liquid:
```diff
@@ -0,0 +1 @@
+{{content}}
```
Since 11ty is a lot less opinionated, it treats a lot more files like templates. Since the default layout is the normal page setup, a bunch of files that shouldn't end up as pages were being treated as other pages. This was the easiest way to add a layout for files that aren't pages.

I'm hoping to find a better way to configure this.

khanhnwin commented on src/_includes/get-sdk.md:
```diff
@@ -13,54 +13,62 @@ which includes the full Dart SDK.
   the Dart SDK.
 
-{{site.alert.important}}
+  :::important
```
love this new syntax!

khanhnwin commented on src/_includes/yt_shims.liquid:
```diff
@@ -1,6 +0,0 @@
-{% assign yt-set = 'frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen loading="lazy"' %}
```
do we define these slugs anywhere else?

khanhnwin replied:
Answered my own question in yt.yml :)

parlough replied:
Not in the same way. These were error-prone as it wasn't clear if you forgot to include this file.

Now they're declared as global data in `src/_data/yt.yml`. You can access any of those properties like `{{yt.embed}}` or `{{yt.std-size}}`.

khanhnwin commented on src/_includes/tools/debug-prod-js-code.md:
```diff
@@ -54,12 +54,12 @@ To debug in Firefox:
 3. Reload the app. The **Debugger** tab shows source-mapped files.
 
-### Safari {#dart2js-debugging-safari}
+### Safari {:#dart2js-debugging-safari}
```
What's the respon for switching this to : as opposed to the standard #?

parlough replied:
Patrice asked the same question :D. For one was consistency with the style we used when adding classes and IDs to elements, but I think I had another reason at the time. I just forgot what that reason was... It was maybe to avoid potential conflicts with another syntax or something?

I'll try to remember and if I can, document it, otherwise consider going back to the shorter syntax consistently.

khanhnwin commented on src/_11ty/plugins/highlight.js:
```diff
@@ -0,0 +1,579 @@
+}
+
+/**
+ * Searches through the specified {@link text} and finds all instances of
```
Love the code comments!

khanhnwin commented on src/_11ty/syntax/dash-light.js:
```diff
@@ -0,0 +1,137 @@
+export default {
+  name: 'dash-light',
```
This is the theme for the site, right? Could use just a line or two to explain it!

parlough replied:
This was originally a JSON file so I couldn't easily add comments. I didn't think to add any when I converted it to a JS file. Good thinking 🧠

khanhnwin commented on src/_11ty/utils/get-page-info.js:
```diff
@@ -0,0 +1,16 @@
+}
+
+// TODO(parlough): Replace this functionality or document it.
+const _cachedPages = new Map();
```
would love to learn how this cache works!

parlough replied:
Me too. I'm hoping I can make it unnecessary, but if not, I'll document it clearly!!

parlough commented on README.md:
```diff
@@ -86,155 +87,144 @@ _choose one_ of the following submodule-cloning techniques:
-Otherwise, you can push your changes.
+If this script reports any errors or warnings,
+then address those issues and rerun the command.
+If you have any issues, reach out for help.
```
I'm not sure how I forgot to include something like this... "reach out for help" is useless :P

I added some more information as well as a link to the Flutter contributors Discord.

Thanks!!!

khanhnwin commented on src/_11ty/filters.js:
```diff
@@ -0,0 +1,182 @@
+  return input.toString().replace(new RegExp(regex), replacement);
+}
+
+export function toISOString(input) {
```
Could we get function comments to explain when/where these filters are used?

parlough replied:
I'd love to. I'd like to do this as follow-up though as I plan to standardize old and new filters.

khanhnwin commented on eleventy.config.js:
```diff
@@ -0,0 +1,135 @@
+import {
```
Perfect, thank you!

khanhnwin commented:
CC @johnpryan you've touched the dart.dev codebase more than I have, check out what Parker's done!!

parlough commented:
> I went through all of the infra changes, mostly glossed over the .md changes as those are just syntax migrations.

Thank you @khanhnwin!! I really appreciate you taking the time and all your feedback. It was super helpful to me. I addressed most of your comments, mentioned them as follow-up, and will work on some final improvements now.

>  (Highlighting code comes to mind!)

Don't worry about this too much as it's the main area I expect to change and be reduced significantly. I'll make sure the final version is easier to read and super well documented.

> We might also want to flesh out the dependency instructions just a tad bit more because my biggest road block was updating node/npm (I think nvm would address this. I thinks contributors should check what they have first as opposed to running into an error and then having to correct it/upgrade. Also enabling corepack (permissions issue, needed sudo).

This is an amazing point. The main goal is to be easy and straightforward to set up so any improvements here are huge. I'm hesitant to suggest `nvm` as that's another dependency on top of what's already needed and it's incompatible with some workflows. In the end, I want the instructions to be agnostic of the installation mechanism, but I'll try some more adjustments to make it clearer, including mentioning a version check.

Could you expand on the issues you had updating node/npm? Or what would have helped you? I'd like to see how I could make it clearer for those that hit similar issues.

> Otherwise, I was up and running super fast! ⚡

Together we can keep making it faster =]

> Thank you thank you thank you!!

Thanks for your help getting to this point. It ended up taking a while to get here as I had to work on other things, so I appreciate your patience!

khanhnwin commented:
Fair point on `nvm`! I hadn't thought of that, but agreed. Standardizing the process should be the goal. 

Here's the steps that I went through:
```
$ pnpm install
ERROR: This version of pnpm requires at least Node.js v16.14
The current version of Node.js is v16.13.0
Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.

$ sudo npm cache clean -f
npm WARN using --force Recommended protections disabled.

 $ sudo npm install -g n
changed 1 package, and audited 2 packages in 630ms
found 0 vulnerabilities

$ sudo n stable
  installing : node-v20.11.0
       mkdir : /usr/local/n/versions/node/20.11.0
       fetch : https://nodejs.org/dist/v20.11.0/node-v20.11.0-darwin-x64.tar.xz
     copying : node/20.11.0
   installed : v20.11.0 (with npm 10.2.4)

$ pnpm install
Lockfile is up to date, resolution step is skipped
Packages: +755
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 8.14.3 → 8.15.1.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v8.15.1   │
   │     Run "corepack prepare pnpm@8.15.1 --activate" to update.     │
   │                                                                  │
   │      Follow @pnpmjs for updates: https://twitter.com/pnpmjs      │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯

Progress: resolved 755, reused 0, downloaded 754, added 755, done
node_modules/.pnpm/protobufjs@7.2.4/node_modules/protobufjs: Running postinstall script, done in 628ms
node_modules/.pnpm/re2@1.20.9/node_modules/re2: Running install script, done in 2.4s

dependencies:
+ bootstrap-scss 4.6.2

devDependencies:
+ @11ty/eleventy 3.0.0-alpha.4
+ firebase-tools 13.0.3
+ hast-util-from-html 2.0.1
+ hast-util-select 6.0.2
+ hast-util-to-html 9.0.0
+ hast-util-to-text 4.0.0
+ html-minifier-terser 7.2.0
+ js-yaml 4.1.0
+ markdown-it 14.0.0
+ markdown-it-anchor 8.6.7
+ markdown-it-attrs 4.1.6
+ markdown-it-container 4.0.0
+ markdown-it-deflist 3.0.0
+ markdown-it-table 4.1.1
+ sass 1.70.0
+ shikiji 0.10.0

Done in 15.2s

```

I'd probably make the first step in the instructions `run node --version  and make sure it is up to date with X version that pnpm requires` - this makes sure that the result of the very first command contributors are running to contribute to the site isn't an error. 

Admittedly it's been a bit since I've worked with node and npm, so it took a google search or two to remember how to upgrade node from the command line instead of downloading it from the website since I was already in the command line, didn't want to go to the browser, download, install, etc.

parlough commented:
Thanks for expanding @khanhnwin!

It's a bit of challenge to resolve all of these potential issues since it really depends on your OS and the mechanism you used to install Node, as well as how you configured it during installation. I've however added some more details, including version checks like you suggested, a note about nvm, and for next steps if you do need to update. Thanks for thinking of this!

I hope to add some extra checks to `dart_site` tool to verify the setup as well. That will point out to users if there are any issues and what steps to consider to resolve them.

atsansone commented:
@parlough : None of the code blocks have the copy button anymore. Will this be redressed in a subsequent PR? If so, no problem. If not, can you create an issue to handle this? I think this functionality is definitely needed.

parlough commented:
@atsansone The Dart website never had a copy button actually, but the changes here will enable me adding one in a way that's consistent and works for here and the Flutter site. It's definitely part of my earlier follow-up work :) - I'll make sure to ping you on that.

chalin commented:
> 👋 Hey Patrice @chalin! I really appreciate you taking a look! I hope you are doing well :)

Hi! 👋🏻 @parlough 

> I'm not sure I would say phenomenal with some of my 11ty and highlighting implementation 🙈, but I'm glad you think so! JS is at least a lot more familiar to all of us than Ruby.

Yes, +100 for JS.

> For more context: The code block functionality ended up being the bulk of my mess and change here. 

👍🏻 

> That's the goal!! Ruby, Jekyll, and Docker or some combination of them have been causing issues often and I think Googlers can't easily use Docker on their macs anymore.

The setup proposed here is sooo much better.

I understand your hesitation to bring in another dependency. But for those who can use it, `nvm` makes installing and periodically updating Node a simple single-command affair :).  Here's how I [phrased the tool-use suggestion elsewhere](https://github.com/open-telemetry/opentelemetry.io/blob/main/CONTRIBUTING.md#local-setup):

> **Install or upgrade** to the [**active LTS** release][nodejs-rel] of **Node.js**. We recommend using **[nvm][]** to manage your Node installation. Under Linux, run the following command (which will install/upgrade to the version specified in [.nvmrc][]):
> 
> ```sh
> nvm install
> ```

That's it, whether I'm installing afresh or wanting to upgrade Node to the latest active LTS release.

[.nvmrc]: https://github.com/open-telemetry/opentelemetry.io/blob/main/.nvmrc
[nvm]:
  https://github.com/nvm-sh/nvm/blob/master/README.md#installing-and-updating
[nodejs-rel]: https://nodejs.org/en/about/previous-releases

> They're mostly the same as you left them, outside of a few fixes and being invoked through that Dart script you mentioned. The changes in this PR are about not using the instructions in code block generation, so handling all Markdown code blocks the same.

👍🏻 

> Based on how we've ended up using the tools though, I started working on a version of the tooling with less functionality that doesn't use `package:build_runner` in [dart-lang/site-shared#201](https://github.com/dart-lang/site-shared/pull/201). It's not done and needs to be tested, but I'd like to get back to it soon :D. If you're interested, I'd appreciate any feedback you have on that!

I took a quick look. I think that it's a great idea to avoid `build_runner`. Your proposed simplifications seem quite reasonable to me; the original tool had features inherited from the Angular Dart and Angular JS days.

I'd love to use such a "lighter-weight" tool. Looking forward to see [dart-lang/site-shared#201](https://github.com/dart-lang/site-shared/pull/201) land as well. (I've subscribed so that I can keep up with your work.)

> There aren't too many usages yet, but there's a few in [`/effective-dart/usage#prefer-relative-import-paths`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/effective-dart/usage#prefer-relative-import-paths). Each linter rule page uses them as well: [`/tools/linter-rules/avoid_relative_lib_imports#usage`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/tools/linter-rules/avoid_relative_lib_imports#usage).
> 
> For line numbers, I think [`/tools/analysis`](https://dart-dev--pr5483-feat-eleventy-migration-ph0qb4b2.web.app/tools/analysis) might be the only usage so far.

Thanks for pointing out those examples.

> A side note, thanks for your documentation and examples on updating Docsy and its downstream sites to Bootstrap 5! I really need to work on that migration soon and your work will be a super useful reference.

I was wondering about that (when I saw the BS v4) dependency still there. I'll be very glad if any of that is helpful to others. Let me know if you have any questions. (Interestingly, I might be tackling the Docsy upgrade of BS from 5.2 to 5.3 next week. It isn't a major version upgrade, but there are _significant_ changes in the way colors are handled.)

Again, great work. I'm glad to see this merged ✨
